cartridge/shop/views.py

```diff
@@ -282,6 +299,15 @@ def checkout_steps(request, form_class=OrderForm, extra_context=None):
                 # and send the order receipt email.
                 order = form.save(commit=False)
                 order.setup(request)
+
+                if (not settings.SHOP_PAYMENT_STEP_ENABLED or
+                        order.total == 0):
```

Comment:
The first condition `not settings.SHOP_PAYMENT_STEP_ENABLE` is to handle the issue I mentioned, and the second `order.total == 0` is to handle the original problem.

These conditions will be evaluated during the very last step of the checkout, so these are possible. step += 1 just skip the payment form page.

Sorry if it isn't clear, this problem isn't as easy to fix as I initially though it would be.